models.js

```diff
@@ -15,6 +15,8 @@ debuglog = function () {
 
 (function() {
     var xpm = xpathmodels;
+    xpm.validHashtagNamespaces = [];
+    xpm.hashtagToXPathDict = {};
```

Comment:
Since `xpathmodels` (`xpm` here) is a global singleton, I don't think it's a good idea to add mutable state like this. If we want the parser to be stateful, then we should implement that as non-global state with an API something like:
```js
var xpathmodels = XPathModels();
xpathmodels.setWhatever(...)
```
Also, passing in valid namespaces like this makes it impossible to parse an expression that may have arbitrary namespaces that are not known at parse time. The same goes for hashtag to xpath transformations (`hashtagToXPathDict`). I think it may be better to pass in an object that knows how to validate hashtag namespaces and convert hashtags to xpath expressions. Here are three different implementation examples to give you an idea of what I'm thinking of:
```js
// reject all hashtags
var hashtags = {
        isValidNamespace: function (value) {
            return false;
        }
        toXpath: function (hashtagExpr) {
            throw new Error("not supported");
        }
    },
    xpathmodels = XPathModels();
xpathmodels.setHashtagInterpreter(hashtags);
```
```js
// convert hashtags in known namespaces to xpath expressions
function makeKnownHashtagInterpreter(namespaces, xpaths) {
    namespaces = _.object(_.map(namespaces, function (value) {
        return [value, null];
    }));
    return {
        isValidNamespace: function (value) {
            return namespaces.hasOwnProperty(value);
        },
        toXpath: function (hashtagExpr) {
            return xpaths[hashtagExpr.toHashtag()];
        },
    }
}

var hashtags = makeHashtagInterpreter(
        ["form"],
        {"#form/question": "/data/question"}
    ),
    xpathmodels = XPathModels();
xpathmodels.setHashtagInterpreter(hashtags);
```
```js
// process arbitrary hashtags with passthrough conversion
function makePassthroughHashtagInterpreter() {
    return {
        isValidNamespace: function (value) {
            return true;
        },
        toXpath: function (hashtagExpr) {
            return hashtagExpr.toHashtag();
        },
    }
}

var hashtags = makeHashtagInterpreter(),
    xpathmodels = XPathModels();
xpathmodels.setHashtagInterpreter(hashtags);
```